Media control time fields should be Locale-aware

jernoble opened the pull request:
#### e23794f045e76d54f592ca67e57c5ad481bed967
<pre>
Media control time fields should be Locale-aware
<a href="https://rdar.apple.com/141281469">rdar://141281469</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296748">https://bugs.webkit.org/show_bug.cgi?id=296748</a>

Reviewed by Andy Estes.

The current time, elapsed time, and duration fields in the UA media controls currently are
hard-coded to a western digital time format, regardless of the number format for the user&apos;s
preferred locale.

Adopt Int.DurationFormat() as the mechanism for formatting the duration strings in those time
fields. Because the width of those fields is no longer fixed and statically calulatable, the layout
of the TimeControl object must be made more flexable.

Update the TimeContol object to use flex-box to layout the elapsedTime, scrubber, remainingTime, and
duration objects. Use CSS rather than JavaScript to control object visibility. Add container layout
style rules to dynamically hide and show time labels as the TimeControl object resizes.

* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-layout-expected.txt:
* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-layout.html:
* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-time-control-styles-expected.txt:
* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-time-control-styles.html:
* LayoutTests/media/modern-media-controls/resources/media-controls-loader.js:
* LayoutTests/media/modern-media-controls/time-control/time-control-expected.txt:
* LayoutTests/media/modern-media-controls/time-control/time-control.html:
* LayoutTests/media/modern-media-controls/time-label/ios-time-label-expected.txt:
* LayoutTests/media/modern-media-controls/time-label/ios-time-label.html:
* LayoutTests/media/modern-media-controls/time-label/time-label-expected.txt:
* LayoutTests/media/modern-media-controls/time-label/time-label.html:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/modern-media-controls/controls/activity-indicator.css:
(.activity-indicator):
* Source/WebCore/Modules/modern-media-controls/controls/slider.css:
(.slider.default &gt; .appearance &gt; .fill):
(.slider.default &gt; .appearance &gt; .fill &gt; :first-child):
(.slider.default &gt; .appearance &gt; .fill &gt; :last-child):
(.slider.default.allows-relative-scrubbing &gt; .appearance.changing &gt; .fill):
(.slider.default &gt; .appearance &gt; .fill &gt; *):
(.slider.default &gt; .appearance &gt; .fill &gt; .primary):
(.slider.default &gt; .appearance &gt; .fill &gt; .track):
(.slider.default.allows-relative-scrubbing &gt; .appearance &gt; .fill &gt; .track &gt; .secondary):
(.slider.default &gt; .appearance &gt; .fill &gt; .knob):
(.slider.default &gt; .appearance &gt; .fill &gt; .knob.circle):
(.slider.default &gt; .appearance &gt; .fill &gt; .knob.bar):
(.slider.default &gt; .appearance &gt; .fill &gt; .knob.none):
(.slider.default.disabled &gt; .appearance &gt; .fill &gt; :is(.primary, .secondary)):
(.slider.default.allows-relative-scrubbing &gt; .appearance &gt; .fill &gt; .secondary): Deleted.
(.slider.default &gt; .appearance &gt; .knob): Deleted.
(.slider.default &gt; .appearance &gt; .knob.circle): Deleted.
(.slider.default &gt; .appearance &gt; .knob.bar): Deleted.
(.slider.default &gt; .appearance &gt; .knob.none): Deleted.
(.slider.default.disabled &gt; .appearance &gt; .fill &gt; :is(.primary, .secondary),): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/slider.js:
(Slider.prototype.commit):
* Source/WebCore/Modules/modern-media-controls/controls/time-control.css: Copied from Source/WebCore/Modules/modern-media-controls/controls/time-label.css.
(.time-control):
(.time-control .scrubber):
(.time-control.above,):
(.time-control.above):
(.time-label):
(.time-control.below):
(.time-control.duration &gt; #time-label-remaining):
(.time-control.remaining &gt; #time-label-duration):
(@container time-control-container (width &lt; 160px)):
* Source/WebCore/Modules/modern-media-controls/controls/time-control.js:
(TimeControl.prototype.this.remainingTimeLabel):
(TimeControl.prototype.get minimumWidth):
(TimeControl.prototype.get idealMinimumWidth):
(TimeControl.prototype.layout):
(TimeControl.prototype.handleEvent):
(TimeControl.prototype._performIdealLayout):
(TimeControl.prototype._toggleDurationRemainingLabel):
(TimeControl.): Deleted.
(TimeControl.prototype.get timeLabelsAttachment): Deleted.
(TimeControl.prototype.set timeLabelsAttachment): Deleted.
(TimeControl.prototype.get _timeLabelsDisplayOnScrubberSide): Deleted.
(TimeControl.prototype.get _canShowDurationTimeLabel): Deleted.
(TimeControl.prototype._durationOrRemainingTimeLabel): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/time-label.css:
(.time-label):
* Source/WebCore/Modules/modern-media-controls/controls/time-label.js:
(TimeLabel.prototype.set value):
(TimeLabel.prototype.get numberOfDigits):
(TimeLabel.prototype.set numberOfDigits):
(TimeLabel.prototype.setValueWithNumberOfDigits):
(TimeLabel.prototype._formattedTime):
* Source/WebCore/Modules/modern-media-controls/controls/tvos-media-controls.js:
* Source/WebCore/Modules/modern-media-controls/controls/vision-media-controls.js:
(VisionInlineMediaControls.prototype._createControls):
(VisionInlineMediaControls.prototype._performSingleBarLayout):
(VisionInlineMediaControls):
* Source/WebCore/Modules/modern-media-controls/main.js:
(UIString):
(formatTimeByUnit): Deleted.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/299348@main">https://commits.webkit.org/299348@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b461b6fa5c07ac8fb7cd6b9fc3b83ab8a89c9153

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27688 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123499 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69388 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4bf6ff75-ebb5-42d6-9104-b5fb06746c80) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119278 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37766 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89086 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43750 "Too many flaky failures: css3/filters/drop-shadow-current-color.html, fast/canvas/offscreen-no-script-context-crash.html, http/tests/resourceLoadStatistics/only-accept-first-party-cookies-with-third-party-cookie-blocking.html, imported/w3c/web-platform-tests/event-timing/contextmenu.html, imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative.html, imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative.html, js/ToNumber.html, js/dom/eval-contained-syntax-error.html, js/dom/webassembly-memory-shared-fail.html, storage/indexeddb/exception-in-event-aborts-private.html, storage/indexeddb/modern/get-index-failures.html, webgl/2.0.0/conformance2/textures/canvas_sub_rectangle/tex-2d-rg16f-rg-float.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_blob/tex-3d-srgb8_alpha8-rgba-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_canvas/tex-2d-rgb5_a1-rgba-unsigned_byte.html, webgl/2.0.y/conformance/more/functions/vertexAttribPointerBadArgs.html (failure)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c8810ca0-8360-4fb7-8db9-af9a12780b09) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120352 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30054 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69596 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c336d470-b581-4a66-b9bf-01a909d77466) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29111 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23377 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67172 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99457 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126620 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44297 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33289 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97752 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44655 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97546 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25046 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42916 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20851 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40647 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44170 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49829 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43627 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46971 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45322 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b461b6fa5c07ac8fb7cd6b9fc3b83ab8a89c9153. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48778

jernoble committed:
Media control time fields should be Locale-aware
rdar://141281469
https://bugs.webkit.org/show_bug.cgi?id=296748

Reviewed by Andy Estes.

The current time, elapsed time, and duration fields in the UA media controls currently are
hard-coded to a western digital time format, regardless of the number format for the user's
preferred locale.

Adopt Int.DurationFormat() as the mechanism for formatting the duration strings in those time
fields. Because the width of those fields is no longer fixed and statically calulatable, the layout
of the TimeControl object must be made more flexable.

Update the TimeContol object to use flex-box to layout the elapsedTime, scrubber, remainingTime, and
duration objects. Use CSS rather than JavaScript to control object visibility. Add container layout
style rules to dynamically hide and show time labels as the TimeControl object resizes.

* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-layout-expected.txt:
* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-layout.html:
* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-time-control-styles-expected.txt:
* LayoutTests/media/modern-media-controls/macos-inline-media-controls/macos-inline-media-controls-time-control-styles.html:
* LayoutTests/media/modern-media-controls/resources/media-controls-loader.js:
* LayoutTests/media/modern-media-controls/time-control/time-control-expected.txt:
* LayoutTests/media/modern-media-controls/time-control/time-control.html:
* LayoutTests/media/modern-media-controls/time-label/ios-time-label-expected.txt:
* LayoutTests/media/modern-media-controls/time-label/ios-time-label.html:
* LayoutTests/media/modern-media-controls/time-label/time-label-expected.txt:
* LayoutTests/media/modern-media-controls/time-label/time-label.html:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/modern-media-controls/controls/activity-indicator.css:
(.activity-indicator):
* Source/WebCore/Modules/modern-media-controls/controls/slider.css:
(.slider.default > .appearance > .fill):
(.slider.default > .appearance > .fill > :first-child):
(.slider.default > .appearance > .fill > :last-child):
(.slider.default.allows-relative-scrubbing > .appearance.changing > .fill):
(.slider.default > .appearance > .fill > *):
(.slider.default > .appearance > .fill > .primary):
(.slider.default > .appearance > .fill > .track):
(.slider.default.allows-relative-scrubbing > .appearance > .fill > .track > .secondary):
(.slider.default > .appearance > .fill > .knob):
(.slider.default > .appearance > .fill > .knob.circle):
(.slider.default > .appearance > .fill > .knob.bar):
(.slider.default > .appearance > .fill > .knob.none):
(.slider.default.disabled > .appearance > .fill > :is(.primary, .secondary)):
(.slider.default.allows-relative-scrubbing > .appearance > .fill > .secondary): Deleted.
(.slider.default > .appearance > .knob): Deleted.
(.slider.default > .appearance > .knob.circle): Deleted.
(.slider.default > .appearance > .knob.bar): Deleted.
(.slider.default > .appearance > .knob.none): Deleted.
(.slider.default.disabled > .appearance > .fill > :is(.primary, .secondary),): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/slider.js:
(Slider.prototype.commit):
* Source/WebCore/Modules/modern-media-controls/controls/time-control.css: Copied from Source/WebCore/Modules/modern-media-controls/controls/time-label.css.
(.time-control):
(.time-control .scrubber):
(.time-control.above,):
(.time-control.above):
(.time-label):
(.time-control.below):
(.time-control.duration > #time-label-remaining):
(.time-control.remaining > #time-label-duration):
(@container time-control-container (width < 160px)):
* Source/WebCore/Modules/modern-media-controls/controls/time-control.js:
(TimeControl.prototype.this.remainingTimeLabel):
(TimeControl.prototype.get minimumWidth):
(TimeControl.prototype.get idealMinimumWidth):
(TimeControl.prototype.layout):
(TimeControl.prototype.handleEvent):
(TimeControl.prototype._performIdealLayout):
(TimeControl.prototype._toggleDurationRemainingLabel):
(TimeControl.): Deleted.
(TimeControl.prototype.get timeLabelsAttachment): Deleted.
(TimeControl.prototype.set timeLabelsAttachment): Deleted.
(TimeControl.prototype.get _timeLabelsDisplayOnScrubberSide): Deleted.
(TimeControl.prototype.get _canShowDurationTimeLabel): Deleted.
(TimeControl.prototype._durationOrRemainingTimeLabel): Deleted.
* Source/WebCore/Modules/modern-media-controls/controls/time-label.css:
(.time-label):
* Source/WebCore/Modules/modern-media-controls/controls/time-label.js:
(TimeLabel.prototype.set value):
(TimeLabel.prototype.get numberOfDigits):
(TimeLabel.prototype.set numberOfDigits):
(TimeLabel.prototype.setValueWithNumberOfDigits):
(TimeLabel.prototype._formattedTime):
* Source/WebCore/Modules/modern-media-controls/controls/tvos-media-controls.js:
* Source/WebCore/Modules/modern-media-controls/controls/vision-media-controls.js:
(VisionInlineMediaControls.prototype._createControls):
(VisionInlineMediaControls.prototype._performSingleBarLayout):
(VisionInlineMediaControls):
* Source/WebCore/Modules/modern-media-controls/main.js:
(UIString):
(formatTimeByUnit): Deleted.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/299348@main
